### PR TITLE
Add python to docs

### PR DESCRIPTION
**Pull request by SimonWoolf**

### CODE REVIEW

File: content/rest/_stats.textile [hunk 1]
```diff
@@ -9,8 +9,8 @@ h4. <span lang="default">@options@ parameters</span><span lang="objc,swift">@ART
 
 The following options, as defined in the "REST @/stats@ API":/rest-api#stats endpoint, are permitted:
 
-- <span lang="ruby">:</span>start := _beginning of time_ earliest <span lang="ruby">@Time@ or </span>millisecond since epoch time for any stats retrieved<br>__Type: @Long@<span lang="ruby">, @Time@</span>__
-- <span lang="ruby">:</span>end := _current time_ latest <span lang="ruby">@Time@ or </span>millisecond since epoch time for any stats retrieved<br>__Type: @Long@<span lang="ruby">, @Time@</span>__
+- <span lang="ruby">:</span>start := _beginning of time_ earliest <span lang="ruby">@Time@ or </span>millisecond since epoch time for any stats retrieved<br>__Type: <span lang="default">@Integer@</span><span lang="java">@Long Integer@</span><span lang="ruby">@Time@</span><span lang="objc,swift">@NSDate@</span>__
```
Comment (mattheworiordan) [inline content/rest/_stats.textile]: millsecond**s** since epoch

File: content/rest/authentication.textile [hunk 2]
```diff
@@ -346,9 +368,12 @@ bc[java]. try {
   System.out.println("An error occurred; err = " + e.getMessage());
 }
 
-bc[ruby]. token_request = client.auth.create_token_request(client_id: 'bob')
+bc[ruby]. token_request = client.auth.create_token_request(c#{lient_id: 'bob')
```
Comment (mattheworiordan) [inline content/rest/authentication.textile]: Typo here

File: content/rest/channels-messages.textile [hunk 3]
```diff
@@ -52,6 +52,14 @@ The Ably REST client library provides a straightforward API for publishing and r
   puts "Last published message: #{result_page.items.first}"
 ```
 
+```[python]
+  rest = AblyRest('{{API_KEY}}')
+  channel = rest.channels.get('{{RANDOM_CHANNEL_NAME}}')
+  channel.publish(u'example', u'message data')
```
Comment (mattheworiordan) [inline content/rest/channels-messages.textile]: Out of interest, why the `u` prefix on both?

File: content/rest/channels-messages.textile [hunk 4]
```diff
@@ -336,13 +361,11 @@ Gets a "paginated":#paginated-result set of historical messages for this channel
 
 h4. Parameters
 
-- <span lang="default">options</span><span lang="objc,swift">query</span><span lang="java">"Param":#param[] options</span> := <span lang="default">an optional object containing the query parameters</span><span lang="ruby">an optional set of key value pairs containing the query parameters</span><span lang="ruby">an optional Associative Array containing the query parameters</span>, as specified in the "message history API documentation":/realtime/history#channel-history.
+- <span lang="default">options</span><span lang="objc,swift">query</span><span lang="java">"Param":#param[] options</span> := <span lang="default">an optional object containing the query parameters</span><span lang="python">optional keyword arguments containing the query parameters</span><span lang="ruby">an optional set of key value pairs containing the query parameters</span><span lang="ruby">an optional Associative Array containing the query parameters</span>, as specified in the "message history API documentation":/rest/history#channel-history.
```
Comment (mattheworiordan) [inline content/rest/channels-messages.textile]: This is wrong, although was wrong before, the last span should be PHP

File: content/rest/channels-messages.textile [hunk 5]
```diff
@@ -336,13 +361,11 @@ Gets a "paginated":#paginated-result set of historical messages for this channel
...
 
 - <div lang="jsall">callback</div> := is a function of the form: @function(err, resultPage)@
 - <div lang="swift,objc">callback</div> := called with a "ARTPaginatedResult":#paginated-result<"ARTMessage":#message> object or an error
 
-Further details of the supported @options@ params, see "message history API documentation":/rest/history#channel-history.
```
Comment (mattheworiordan) [inline content/rest/channels-messages.textile]: Why did you remove this?

File: content/rest/channels-messages.textile [hunk 6]
```diff
@@ -209,6 +224,12 @@ if result_page.has_next?
   puts "#{next_page.items.length} messages received in second page"
 end
 
+bc[python]. result_page = channel.history()
+print str(len(result_page.items)) + " messages received in first page"
```
Comment (mattheworiordan) [inline content/rest/channels-messages.textile]: You've used single quotes elsewhere, any reason for doubles here and below?

File: content/rest/history.textile [hunk 7]
```diff
@@ -167,15 +176,15 @@ Gets a "paginated":#paginated-result set of historical presence events for this
...
 
 h4. <span lang="default">@options@ parameters</span><span lang="objc,swift">@ARTDataQuery@ properties</span>
 
-- <span lang="ruby">:</span>start := _beginning of time_ earliest <span lang="ruby">@Time@ or </span>millisecond since epoch time for any presence event retrieved<br>__Type: @Long@<span lang="ruby">, @Time@</span>__
-- <span lang="ruby">:</span>end := _current time_ latest <span lang="ruby">@Time@ or </span>millisecond since epoch time for any presence events retrieved<br>__Type: @Long@<span lang="ruby">, @Time@</span>__
+- <span lang="ruby">:</span>start := _beginning of time_ earliest <span lang="ruby">@Time@ or </span>millisecond since epoch time for any presence event retrieved<br>__Type: <span lang="default">@Integer@</span><span lang="java">@Long Integer@</span><span lang="ruby">@Time@</span><span lang="objc,swift">@NSDate@</span>__
+- <span lang="ruby">:</span>end := _current time_ latest <span lang="ruby">@Time@ or </span>millisecond since epoch time for any presence events retrieved<br>__Type: <span lang="default">@Integer@</span><span lang="java">@Long Integer@</span><span lang="ruby">@Time@</span><span lang="objc,swift">@NSDate@</span>__
```
Comment (mattheworiordan) [inline content/rest/history.textile]: milliseconds since epoch, prob worth a global check

File: content/rest/presence.textile [hunk 8]
```diff
@@ -56,6 +56,13 @@ if members_page.has_next?
 end
 ```
 
+```[python]
+members_page = channel.presence.get()
+print str(len(members_page.items)) + " members present"
```
Comment (mattheworiordan) [inline content/rest/presence.textile]: Double quotes here agian

File: content/rest/presence.textile [hunk 9]
```diff
@@ -193,15 +208,15 @@ Gets an array of members present on the channel as "PresenceMessage":#presence-m
 
 h4. Parameters
 
-- <span lang="default">options</span><span lang="objc,swift">query</span><span lang="java">"Param":#param[] options</span><span lang="ruby">options</span> := <span lang="default">an optional object containing query parameters</span><span lang="ruby">an optional set of symbol key and value pairs containing the query parameters</span><span lang="php">an optional Associate Array containing the query parameters</span> as specified below.
+- <span lang="default">options</span><span lang="objc,swift">query</span><span lang="java">"Param":#param[] options</span><span lang="ruby">options</span> := <span lang="default">an optional object containing query parameters</span><span lang="python">optional keyword arguments containing the query parameters</span><span lang="ruby">an optional set of symbol key and value pairs containing the query parameters</span><span lang="php">an optional Associate Array containing the query parameters</span> as specified below.
```
Comment (mattheworiordan) [inline content/rest/presence.textile]: No need for `lang=ruby options` as it's the same as the default

File: content/rest/types.textile [hunk 10]
```diff
@@ -36,22 +36,24 @@ h2. REST Data types
...
-  java:    io.ably.lib.types.ErrorInfo
-  ruby:    Ably::Models::ErrorInfo
-  php:     Ably\Models\ErrorInfo
+blang[jsall,ruby,php,java,objc,swift].
+  h3(#error-info).
+    default: ErrorInfo
+    objc,swift: ARTErrorInfo
+    java:    io.ably.lib.types.ErrorInfo
+    ruby:    Ably::Models::ErrorInfo
+    php:     Ably\Models\ErrorInfo
```
Comment (mattheworiordan) [inline content/rest/types.textile]: That's odd, so no ErrorInfo type in Python?

File: content/rest/usage.textile [hunk 11]
```diff
@@ -72,6 +72,24 @@ blang[ruby].
...
+  Then it can be imported and instantiated as follows:
+
+  ```[python]
+  from ably import AblyRest
+  client = AblyRest(api_key)
+  ```
+
+  h4. Note on string encodings
+
+  Since Ably supports both string and binary payloads, to avoid ambiguity, we recommend that strings passed to the library for publishing to Ably (eg as an event name or payload data) should be unicode strings. In Python 3 this is the normal string type, but in Python 2 it is not, so we suggest you prefix string literals with @u@ prefix (eg @u'eventname'@ - or alternatively, use @from __future__ import unicode_literals@, which will make this automatic), and to explicitly decode any user input (eg @raw_input().decode(sys.stdin.encoding@).
```
Comment (mattheworiordan) [inline content/rest/usage.textile]: Ignore my earlier question re `u`, this answers it :)

File: content/rest/usage.textile [hunk 12]
```diff
@@ -203,6 +223,7 @@ A reference to the "@Channel@ collection":/realtime/channels-messages instance f
 h6(#client-id).
   default: clientId
   ruby:    client_id
+  python:    client_id
```
Comment (mattheworiordan) [inline content/rest/usage.textile]: why not match the indentation?

File: content/types/_client_options.textile [hunk 13]
```diff
@@ -14,28 +14,29 @@ h4.
...
+- <span lang="default">clientId</span><span lang="python">client_id</span><span lang="ruby">:client_id</span> := A client ID, used for identifying this client when publishing messages or for presence purposes. The <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> can be any non-empty string. This option is primarily intended to be used in situations where the library is instanced with a key; note that a <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> may also be implicit in a token used to instance the library; an error will be raised if a <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> specified here conflicts with the <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> implicit in the token. "Find out more about client identities":/how-ably-works#client-identity<br>__Type: @String@__
```
Comment (mattheworiordan) [inline content/types/_client_options.textile]: missing python versions of clientId in the spans below, I expect you simply need to change `lang="ruby"` to `lang="ruby,python"` in most cases

File: content/types/_client_options.textile [hunk 14]
```diff
@@ -14,28 +14,29 @@ h4.
...
+- <span lang="default">authCallback</span><span lang="python">auth_callback</span><span lang="ruby">:auth_callback</span> := A <span lang="default">function</span><span lang="java">@TokenCallback@ instance</span><span lang="python">lambda</span><span lang="ruby">proc / lambda</span> to call when a new token is required. The role of the callback is to either generate a signed "@TokenRequest@":/realtime/types#token-request which may then be submitted automatically by the library to the "Ably REST API @requestToken@":/rest-api#request-token; or to provide a valid token in as a "@TokenDetails@":/realtime/types#token-details object. See "authentication":/rest/authentication for details of the token request format and associated API calls.<br>__Type: <span lang="default">@Function@</span><span lang="java">@TokenCallback@</span><span lang="ruby">@Proc@</span>__
```
Comment (mattheworiordan) [inline content/types/_client_options.textile]: Is the type `Function` right?

File: content/types/_client_options.textile [hunk 15]
```diff
@@ -14,28 +14,29 @@ h4.
...
 
-- <span lang="default">authHeaders</span><span lang="ruby">:auth_headers</span> := A set of key value pair headers to be added to any request made to the @authUrl@. Useful when an application requires these to be added to validate the request or implement the response.<br>__Type: <span lang="default">@Object@</span><span lang="ruby">@Hash@</span><span lang="php">@Associative Array@</span><span lang="java">@Param []@</span>__
+- <span lang="default">authHeaders</span><span lang="python">auth_headers</span><span lang="ruby">:auth_headers</span> := A set of key value pair headers to be added to any request made to the @authUrl@. Useful when an application requires these to be added to validate the request or implement the response.<br>__Type: <span lang="default">@Object@</span><span lang="ruby">@Hash@</span><span lang="php">@Associative Array@</span><span lang="java">@Param []@</span>__
```
Comment (mattheworiordan) [inline content/types/_client_options.textile]: Think you're missing the type here, I assume it's not an `Object` for Python. Similar examples below

File: content/types/_paginated_result.textile [hunk 16]
```diff
@@ -4,16 +4,19 @@ h4.
   default: Properties
   java:    Members
   ruby:    Attributes
+  python:  Attributes
 
-- items := contains a page of results (for example an Array of "@Message@":#message or "@PresenceMessage@":#presence-message objects for a channel history request)<br>__Type: @Array <Message, Presence, Stats>@__
+- items := contains a page of results (for example an Array of "@Message@":#message or "@PresenceMessage@":#presence-message objects for a channel history request)<br>__Type: @List <Message, Presence, Stats>@__
```
Comment (mattheworiordan) [inline content/types/_paginated_result.textile]: You've just made this a list for all langs, whereas others generally refer to it as an Array?

File: content/types/_paginated_result.textile [hunk 17]
```diff
@@ -92,6 +97,15 @@ channel.history do |first_page|
 end
 ```
 
+```[python]
+result_page = channel.history()
+print "Page 0 item 0: " + str(result_page.items[0].data)
```
Comment (mattheworiordan) [inline content/types/_paginated_result.textile]: Double quotes here again

File: content/types/_token_params.textile [hunk 18]
```diff
@@ -17,13 +20,14 @@ h4.
...
+- <span lang="default">clientId</span><span lang="python">client_id</span><span lang="ruby">:client_id</span> := A client ID, used for identifying this client when publishing messages or for presence purposes. The <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> can be any non-empty string. This option is primarily intended to be used in situations where the library is instanced with a key; note that a <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> may also be implicit in a token used to instance the library; an error will be raised if a <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> specified here conflicts with the <span lang="default">@clientId@</span><span lang="ruby">@client_id@</span> implicit in the token. "Find out more about client identities":/how-ably-works#client-identity<br>__Type: @String@__
```
Comment (mattheworiordan) [inline content/types/_token_params.textile]: You replaced the first span but not the others.

**Comment by mattheworiordan** [comment]
Thanks @SimonWoolf, quite a few small comments, but really good.  Once fixed, we can merge in and I will publish to [www.ably.io/documentation](https://www.ably.io/documentation)

File: content/rest/channels-messages.textile (hunk 5 above)
Comment (SimonWoolf) [inline content/rest/channels-messages.textile]: It just repeats the last sentence of the previous paragraph - "...query parameters as specified in the [message history API documentation](). .. Further details of the supported options params, see [message history API documentation]()"

File: content/rest/channels-messages.textile (hunk 6 above)
Comment (SimonWoolf) [inline content/rest/channels-messages.textile]: No good reason, nope. will change
(my Elixir brain is used to typing double quotes for strings, single quotes in Elixir are a different datatype)

File: content/rest/types.textile (hunk 10 above)
Comment (SimonWoolf) [inline content/rest/types.textile]: Nope - afaict any error response returned by realtime is just raise as an AblyException immediately https://github.com/ably/ably-python/blob/master/ably/util/exceptions.py#L35-L70. Presumably that's a pythonic way of doing it, shrug

File: content/rest/channels-messages.textile (hunk 5 above)
Comment (mattheworiordan) [inline content/rest/channels-messages.textile]: True

File: content/types/_client_options.textile (hunk 14 above)
Comment (SimonWoolf) [inline content/types/_client_options.textile]: On double-checking, apparently in python you can pass either a named function or a lambda, and they can be called the same way. So I think I should probably change 'lambda' to just 'function' at the beginning of the para to just 'function', and assume that python devs will know what I mean

File: content/rest/authentication.textile [hunk 19]
```diff
@@ -373,25 +398,27 @@ client.auth.createTokenRequest(tokenParams, options: nil) { tokenRequest, error
...
+- <span lang="ruby,python">auth_options</span><span lang="default">authOptions</span> := <span lang="default">an optional object containing the "authentication options":#auth-options</span><span lang="python">an optional Dict containing the "authentication options":#auth-options</span><span lang="java">an optional "@TokenParams@":#auth-options object containing the authentication options</span><span lang="ruby">an optional set of key value pairs containing the "authentication options":#auth-options</span><span lang="php">an optional set of key value pairs in an associative array containing the "authentication options":#auth-options</span> for the requested token<br>__Type: "@AuthOptions@":#auth-options__
```
Comment (fjsj) [inline content/rest/authentication.textile]: here in Python auth-options aren't a dict, but a list of parameters

File: content/types/_paginated_result.textile (hunk 16 above)
Comment (SimonWoolf) [inline content/types/_paginated_result.textile]: oops, must have thought I was in a bc[python] block

File: content/rest/encryption.textile [hunk 20]
```diff
@@ -61,6 +67,9 @@ h2(#getting-started). Getting started
 
 Note that the @key@ should not be a passphrase, but a cryptographic key - generated from a secure random source, 128 or 256 bits long, binary or base64-encoded. If you wish to encrypt messages with a passphrase (for example, one entered by a user), you should use a "key derivation function":https://en.wikipedia.org/wiki/Key_derivation_function to transform that into a key. The client libraries are also capable of "generating a random key":#generate-random-key for you.
 
+blang[python].
+  If you are using python 2 and you wish to pass in a base64-encoded key, make sure you pass it in as a @unicode@ string, not a @str@, or the library will interpret it as a binary.
```
Comment (fjsj) [inline content/rest/encryption.textile]: python 2 -> Python 2

File: content/rest/encryption.textile [hunk 21]
```diff
@@ -110,11 +121,11 @@ You will rarely need to call this yourself, since the client library will handle
 
 h4. Parameters
 
-- params := The cipher params that you want to specify. It must at a minimum include a @key@, which should be either a binary (<span lang="java">@byte[]@</span><span lang="javascript">@ArrayBuffer@ or @WordArray@</span><span lang="nodejs">@Buffer@</span><span lang="ruby">byte array</span><span lang="php">binary string</span><span lang="objc,swift">@NSData@</span>) or a base64-encoded @String@.
+- params := The cipher params that you want to specify. It must at a minimum include a @key@, which should be either a binary (<span lang="java">@byte[]@</span><span lang="javascript">@ArrayBuffer@ or @WordArray@</span><span lang="nodejs">@Buffer@</span><span lang="python">@bytearray@ or @bytes@</span><span lang="ruby">byte array</span><span lang="php">binary string</span><span lang="objc,swift">@NSData@</span>) or a base64-encoded @String@.<span lang="python"> If you are using python 2 and you wish to pass in a base64-encoded key, make sure you pass it in as a @unicode@ string, not a @str@, or the library will interpret it as a binary.</span>
```
Comment (fjsj) [inline content/rest/encryption.textile]: python 2 -> Python 2

File: content/rest/history.textile [hunk 22]
```diff
@@ -38,6 +38,13 @@ The Ably Rest client library provides a straightforward API to retrieve "paginat
   puts "Most recent message: #{recent_message.message.id} - #{recent_message.data}")
 ```
 
+```[python]
+  rest = AblyRest('{{API_KEY}}')
+  channel = rest.channels.get('{{RANDOM_CHANNEL_NAME}}')
+  result_page = channel.history()
+  recent_message = result_page.items.first
```
Comment (fjsj) [inline content/rest/history.textile]: `first` should be a method call, i.e., `first()`. Please check

File: content/rest/statistics.textile [hunk 23]
```diff
@@ -37,6 +37,13 @@ The Ably Rest client library provides a straightforward API for retrieving appli
   puts this_hour # => #<Ably::Models::Stat:…
 ```
 
+```[python]
+  rest = AblyRest('{{API_KEY}}')
+  result_page = rest.stats(unit='hour')
+  this_hour = result_page.items.first
```
Comment (fjsj) [inline content/rest/statistics.textile]: `.first` -> `.first()`

File: content/rest/authentication.textile (hunk 19 above)
Comment (SimonWoolf) [inline content/rest/authentication.textile]: True for `request_token` and `create_token_request`, but `authorise` takes it as a dict. (Well, either a dict or an `authOptions`).

(The inconsistency should probably be considered a bug)

File: content/rest/history.textile (hunk 22 above)
Comment (SimonWoolf) [inline content/rest/history.textile]: Oops, failure to reset my brain from ruby / elixir mode :)

File: content/rest/types.textile (hunk 10 above)
Comment (fjsj) [inline content/rest/types.textile]: We discussed this on Slack:
```
Flávio: In fact ErrorInfo is being called AblyException in Python lib. I saw that you generate the same methods of Python lib on Ruby lib. Need to rename AblyException to ErrorInfo?
Matthew: Where an exception is thrown that is one thing, but where an ErrorInfo object is used (less frequently done in REST vs Realtime), then we should return an ErrorInfo object to be consistent.  However, looking through the REST API I can’t see anywhere at present where ErrorInfo is used actually, so perhaps leave as is
```

File: content/rest/usage.textile [hunk 24]
```diff
@@ -232,6 +254,7 @@ h6(#time). time
 bq(definition).
   default: time(callback("ErrorInfo":/realtime/types#error-info err, Number time))
   ruby:    Time time
+  ruby:    Int time
```
Comment (fjsj) [inline content/rest/usage.textile]: ruby two times here, should be python?

File: content/rest/authentication.textile (hunk 19 above)
Comment (fjsj) [inline content/rest/authentication.textile]: rationale: https://github.com/ably/ably-python/pull/34

File: content/types/_client_options.textile (hunk 14 above)
Comment (fjsj) [inline content/types/_client_options.textile]: anything `callable` would work: https://docs.python.org/2.7/library/functions.html#callable

File: content/rest/channels-messages.textile (hunk 6 above)
Comment (fjsj) [inline content/rest/channels-messages.textile]: it is ok in Python to use double quotes for strings that will be outputted and single quotes for internal strings: http://programmers.stackexchange.com/a/155188/80170

File: content/rest/authentication.textile (hunk 19 above)
Comment (fjsj) [inline content/rest/authentication.textile]: Yes, I agree it's better to be consistent

**Comment by SimonWoolf** [comment]
Addressing feedback: 4703f58
